clash-cores/src/Clash/Cores/Xilinx/DcFifo/Explicit.hs

```diff
@@ -0,0 +1,258 @@
+  XilinxFifo
+    { writeReset :: Signal write ResetBusy
+    , isFull :: Signal write Full
+    , writeCount :: Signal write (DataCount depth)
+    , readReset :: Signal read ResetBusy
+    , isEmpty :: Signal read Empty
+    , readCount :: Signal read (DataCount depth)
+    , fifoData :: Signal read (BitVector n)
+    }
+
+-- | Mock Xilinx FIFO, see [documentation](https://docs.xilinx.com/v/u/en-US/pg057-fifo-generator)
+-- for expected behavior.
+--
+-- If 'dcReadDataCount' or 'dcWriteDataCount' are disabled, the relevant signals
+-- will be 'deepErrorX'
+dcFifo ::
+  forall depth n write read .
+  ( KnownNat n
+  , KnownDomain write
+  , KnownDomain read
+
+  , KnownNat depth
+  -- Number of elements should be between [2**4, 2**17] ~ [16, 131072].
+  , 4 <= depth
+  , depth <= 17
+  ) =>
+  DcConfig (SNat depth) ->
+
+  Clock write -> Clock read ->
+  -- | Asynchronous reset
+  Reset read ->
+
+  -- | Write data
+  Signal write (BitVector n) ->
+  -- | Write enable
+  Signal write Bool ->
+  -- | Read enable
+  Signal read Bool ->
+  XilinxFifo read write depth n
+dcFifo dcCfg wClk rClk rst writeData wEnable rEnable =
+  let (wRst, f, wCnt, rRst, e, rCnt, rData) = dcFifo# dcCfg wClk rClk rst writeData wEnable rEnable
+    in XilinxFifo wRst f wCnt rRst e rCnt rData
+
+dcFifo# ::
+  forall depth n write read .
+  ( KnownNat n
+  , KnownDomain write
+  , KnownDomain read
+
+  , KnownNat depth
+  , 4 <= depth
+  , depth <= 17
+  ) =>
+  DcConfig (SNat depth) ->
+
+  Clock write -> Clock read -> Reset read ->
+
+  -- | Write data
+  Signal write (BitVector n) ->
+  -- | Write enable
+  Signal write Bool ->
+  -- | Read enable
+  Signal read Bool ->
+  ( Signal write ResetBusy
+  , Signal write Full
+  , Signal write (DataCount depth)
+
+  , Signal read ResetBusy
+  , Signal read Empty
+  , Signal read (DataCount depth)
+  , Signal read (BitVector n)
+  )
+dcFifo# DcConfig{..} wClk rClk rst writeData wEnable rEnable =
+  let
+    (wRstBusy, wFull, wCnt, rRstBusy, rEmpty, rCnt, rData) = go rstSignalR rstSignalW start wEnable rEnable writeData
+  in
+    ( wRstBusy
+    , wFull
+    , if dcWriteDataCount then wCnt else deepErrorX "Write data count disabled"
+    , rRstBusy
+    , rEmpty
+    , if dcReadDataCount then rCnt else deepErrorX "Read data count disabled"
+    , deepErrorX "No sample" :- rData
+    )
+ where
+  rstSignalR = unsafeToHighPolarity rst
+  rstSignalW = unsafeSynchronizer rClk wClk $ unsafeToHighPolarity rst
+
+  -- reified depth
+  rD = snatToNum @Int (powSNat (SNat @2) (SNat @depth))
+
+  -- TODO: how do we ensure it doesn't "space leak" in Haskell?
+  --
+  -- force slower first?
+  -- https://github.com/clash-lang/clash-compiler/blob/ea114d8edd6a110f72d148203b9db2454cae8f37/clash-prelude/src/Clash/Explicit/BlockRam.hs#L1276-L1280
+
+  go ::
+    Signal read Bool ->
+    Signal write Bool ->
+
+    FifoState n ->
+    Signal write Bool ->
+    Signal read Bool ->
+    Signal write (BitVector n) ->
```

Comment:
Perhaps:

```suggestion
    FifoState n ->
    Signal read Bool -> -- reset
    Signal read Bool -> -- read enabled
    Signal write Bool -> -- reset
    Signal write Bool -> -- write enabled
    Signal write (BitVector n) -> -- write data
```

Just to cluster the read/write domains. This makes it easier to see whether the pattern matches in `go` are correct.

Reply:
Maybe even:


```suggestion
    FifoState n ->
    Signal read Bool -> -- reset
    Signal read Bool -> -- read enabled
    
    Signal write Bool -> -- reset
    Signal write (Maybe (BitVector n)) -> -- write data
```

Reply:
Like the `Bit`/`Bool` thing, I wish we had a good way to deconstruct/construct `Maybe`s in the blackboxes. It would save us a little bit of computation on the Haskell side + annoying wrapper functions + `deepErrorX` everywhere.

Reply:
That would be nice! But would it be a lot of work?